manual/src/cmds/tail-mod-cons.etex

```diff
@@ -0,0 +1,528 @@
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss -> append_flatten xs xss
+
+and[@tail_mod_cons] append_flatten xs xss =
+  match xs with
+  | [] -> flatten xss
+  | x :: xs -> x :: append_flatten xs xss
+\end{caml_example*}
+
+Non-recursive functions can also be annotated "[\@tail_mod_cons]"; this is
+typically useful for local bindings to recursive functions.
+
+Incorrect version:
+\begin{caml_example*}{verbatim}[warning=51,warning=71]
+let[@tail_mod_cons] rec map_vars f exp =
+  let self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+\end{caml_example*}
+
+Recommended fix:
+\begin{caml_example*}{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let[@tail_mod_cons] self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+\end{caml_example*}
+
+In other cases, there is either no benefit in making the called function
+tail-mod-cons, or it is not possible: for example, it is a function
+parameter (the transformation only works with direct calls to
+known functions).
+
+For example, consider a substitution function on binary trees:
+\begin{caml_example*}{verbatim}[warning=72]
+type 'a tree = Leaf of 'a | Node of 'a tree * 'a tree
+
+let[@tail_mod_cons] rec bind (f : 'a -> 'a tree) (t : 'a tree) : 'a tree =
+  match t with
+  | Leaf v -> f v
+  | Node (left, right) ->
+    Node (bind f left, (bind[@tailcall]) f right)
+\end{caml_example*}
+
+Here "f" is a function parameter, not a direct call, and the current
+implementation is strictly first-order, it does not support
+tail-mod-cons arguments. In this case, the user should indicate that
+they realize this call to "f v" is not, in fact, in tail position, by
+using "(f[\@tailcall false]) v".
+
+\begin{caml_example*}{verbatim}
+type 'a tree = Leaf of 'a | Node of 'a tree * 'a tree
+
+let[@tail_mod_cons] rec bind (f : 'a -> 'a tree) (t : 'a tree) : 'a tree =
+  match t with
+  | Leaf v -> (f[@tailcall false]) v
+  | Node (left, right) ->
+    Node (bind f left, (bind[@tailcall]) f right)
+\end{caml_example*}
+
+\section{sec:details}{Details on the transformation}
+
+To use this advanced feature, it helps to be aware that the function transformation produces a specialized function in destination-passing-style.
+
+Recall our "map" example:
+\begin{caml_example*}{verbatim}
+let rec map f l =
+  match l with
+  | [] -> []
+  | x :: xs ->
+    let y = f x in
+    y :: map f xs
+\end{caml_example*}
+
+Below is a description of the transformed program in pseudo-OCaml
+notation: some operations are not expressible in OCaml source code.
+(The transformation in fact happens on the Lambda intermediate
+representation of the OCaml compiler.)
+
+\begin{verbatim}
+let rec map f l =
+  match l with
+  | [] -> []
+  | x :: xs ->
+    let y = f x in
+    let dst = y ::{mutable} Hole in
```

Comment:
I am not completely sure of how I should read `{mutable} Hole`?

Reply:
I guess we could just write `y :: Hole`. My idea with `::{mutable}` was to emphasize that, unlike with normal cons operators, the arguments can be mutated after the fact.

Reply:
I missed the fact that `{mutable}` was a qualifier for `::`. I would propose `::_{mutable}` but that is because I am reading `_` as subscript operator. Maybe  `::[@mutable] Hole`? It is still fantasy syntax but it is clearly an annotation on `::`.